### ISSUE TITLE
Wrapping characters spanning multiple print-width 

### ISSUE THREAD

**Issue by heyrict** [sharkdp/bat#787]
Currently, wrapping will not work as expected if there are characters spanning multiple print-width in a line.

![Screenshot from 2019-12-24 15-42-49](https://user-images.githubusercontent.com/25698503/71401377-9ebd1b80-2664-11ea-9b27-1c3cf57e0d3c.png)

I haven't investigated the code, but I think that probably the print-width of a character is always assumed to be `1` in `bat`. In this case it can be fixed by properly calculating the print-width with some crates e.g. `unicode-width`.

**Comment by sharkdp** [comment]
Thank you for reporting this.

If someone wants to reproduce this:

`test.md`:
```md
Hello, world! Hello, world! Hello, world!
你好, 世界! 你好, 世界! 你好, 世界!
```

``` bash
bat test.md --terminal-width=37
```

**Comment by eth-p** [comment]
This was something I considered when initially adding wrapping support, but I decided not to do it at the time. Different fonts and terminals might affect the character width, and I didn't want to add a feature that assumes all terminal emulators would render unicode similarly to iTerm.

I'm still not sure how it could be implemented to account for that, but some testing with various terminals and symbols would likely be needed.

**Comment by VladimirMarkelov** [comment]
>  Different fonts and terminals might affect the character width

Font does not affect character width (e.g, as far as I know ligatures takes the same space as single characters they consist of). Terminal affect character width slightly: the only thing comes to my mind is terminal CJK mode. Some characters are of 1 or 2 cells width depending on if CJK on or off. As far as I know, the number of those ambiguous characters is very low, and these characters are quite rare. So, it is safe to assume for the first implementation that CJK is always off(or on).

As for other characters width, it is a constant value. To calculate width of a string/character in terminal cells, you can use directly a crate [unicode-width](https://crates.io/crates/unicode-width). Or you may find useful a crate that already solves text wrapping problems: [textwrap](https://crates.io/crates/textwrap)

**Comment by sharkdp** [comment]
Fixed by @Kogia-sima in #815

**Comment by sharkdp** [comment]
This has been fixed in [bat 0.13](https://github.com/sharkdp/bat/releases/tag/v0.13.0).
